ext/semian/tickets.c

```diff
@@ -27,31 +34,93 @@ update_ticket_count(update_ticket_count_t *tc)
 
   return Qnil;
 }
+
 void
-configure_tickets(int sem_id, int tickets, int should_initialize)
+update_tickets_from_quota(int sem_id, double quota)
 {
+  int delta = 0;
+  int tickets = 0;
+  int state = 0;
+  update_ticket_count_t tc;
+
+  // If the configured worker count doesn't match the registered worker count, adjust it.
+  // and adjust the underlying tickets available to match.
+  sem_meta_lock(sem_id);
+  delta = get_sem_val(sem_id, SI_SEM_REGISTERED_WORKERS) - get_sem_val(sem_id, SI_SEM_CONFIGURED_WORKERS);
+  if (delta != 0) {
+    if (perform_semop(sem_id, SI_SEM_CONFIGURED_WORKERS, delta, 0, NULL) == -1) {
+      rb_raise(eInternal, "error setting configured workers, errno: %d (%s)", errno, strerror(errno));
+    }
+
+    // Compute the ticket count
+    tickets = (int) ceil(get_sem_val(sem_id, SI_SEM_CONFIGURED_WORKERS) * quota);
+    tc.sem_id = sem_id;
+    tc.tickets = tickets;
+
+#ifdef DEBUG
+    print_sem_vals(sem_id);
+#endif
+
+    rb_protect((VALUE (*)(VALUE)) update_ticket_count, (VALUE) &tc, &state);
+  }
+  sem_meta_unlock(sem_id);
+
+  if (state) {
+    rb_jump_tag(state);
+  }
+}
+
+void
+configure_tickets(int sem_id, int tickets, double quota, int should_initialize)
+{
+
   if (should_initialize) {
-    initialize_tickets(sem_id, tickets);
+    initialize_tickets(sem_id, tickets, quota);
   }
 
   if (tickets > 0) {
     configure_static_tickets(sem_id, tickets);
+  } else if (quota > 0) {
+    configure_quota_tickets(sem_id, quota);
```

Comment:
Static tickets just sets it directly.

The quota tickets must compute what it *should* be set to first, then calls the same function to update the ticket count.

I'm not quite sure I understand your concern via text